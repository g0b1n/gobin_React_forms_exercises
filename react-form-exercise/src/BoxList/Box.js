import React from "react";

const Box = ({ id, backgroundColor, width, height }) => {
    const boxStyle = {
        backgroundColor,
        width,
        height
    };

    return <div style={boxStyle}></div>
};

export default Box;