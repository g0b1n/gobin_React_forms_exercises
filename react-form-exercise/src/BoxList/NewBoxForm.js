import React, {useState} from "react";

const NewBoxForm = ({addBox}) => {
    const InitialState = {
        backgroundColor: "",
        width: "",
        height: ""
    }

    const [formData, setFormData] = useState(InitialState);
    const handleChange = (e) => {
        const {name, value} = e.target;
        setFormData(formData => ({
            ...formData,
            [name]: value
        }))
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        addBox({...formData})
        setFormData(InitialState)
    }

    

    return (
        <form onSubmit={handleSubmit}>
            <label htmlFor="backgroundColor">Background color</label>
            <input
                id="backgroundColor"
                type="text" name="backgroundColor"
                placeholder="Background Color"
                value={formData.backgroundColor}
                onChange={handleChange}
            />

            <label htmlFor="width">Box width</label>
            <input
                id="width"
                type="text" name="width"
                placeholder="Box width"
                value={formData.width}
                onChange={handleChange}
            />

            <label htmlFor="height">Box height</label>
            <input
                id="height"
                type="text" name="height"
                placeholder="Box height"
                value={formData.height}
                onChange={handleChange}
            />

            <button>Add Box</button>

        </form>
    )
}

export default NewBoxForm;