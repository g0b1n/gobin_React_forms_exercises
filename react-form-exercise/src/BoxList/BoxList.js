import React, {useState} from "react";
import { v4 as uuid } from 'uuid';
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import './BoxList.css';

const BoxList = () => {
    const initial_Box = [
        {id: uuid(), backgroundColor: 'blue', width: '100px', height: '100px'},
        {id: uuid(), backgroundColor: 'red', width: '100px', height: '100px'}
    ]

    const [boxes, setBoxes] = useState(initial_Box)

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id: uuid() }])
    }

    // Delete a box
    const deleteBox = (id) => {
       setBoxes(boxes => boxes.filter(box => box.id !== id));
    }

    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox} />
           {boxes.map(box => (
            <div key={box.id}>
                <Box
                    id={box.id} 
                    backgroundColor={box.backgroundColor} 
                    width={box.width} 
                    height={box.height} 
                />
                <button onClick={ () => deleteBox(box.id)}>X</button>
            </div>
            ))}
        </div>
    )
}

export default BoxList;