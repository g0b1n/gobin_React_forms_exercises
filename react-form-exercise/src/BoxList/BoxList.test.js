import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
test('it renders without crashing', () => {
    render(<BoxList />)
})

// snapshot test
test('it matches the snapshot', () => {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot()
})

