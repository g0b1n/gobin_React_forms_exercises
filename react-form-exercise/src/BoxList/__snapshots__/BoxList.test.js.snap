// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="backgroundColor"
      >
        Background color
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="Background Color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Box width
      </label>
      <input
        id="width"
        name="width"
        placeholder="Box width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Box height
      </label>
      <input
        id="height"
        name="height"
        placeholder="Box height"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
    <div>
      <div
        style="background-color: blue; width: 100px; height: 100px;"
      />
      <button>
        X
      </button>
    </div>
    <div>
      <div
        style="background-color: red; width: 100px; height: 100px;"
      />
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
